### ISSUE TITLE
Use threads instead of Pyro

### ISSUE THREAD

**Issue by ianhi** [tlambert03/pymmcore-plus#70]
From a conversation with @tlambert03:

Pyro was originally introduced in order to do the inter process communication so that actions like `snapImage` weren't blocking. This was bad for napari-micromanager because even in a thread `core.snapImage` was blocking the pyqt event loop. This is because if a thread does not release the GIL then it will block the gui event loop. This was happening even though we were calling out to a C++ function. The reason being that in versions 10.1.1.70.3 and prior of pymmcore the SWIG wrappers weren't built with threading so the GIL was never released (https://github.com/micro-manager/pymmcore/pull/50). 

As a demonstration of version `10.1.1.70.4` no longer having the same blocking behavior run the below example:

note that going live doesn't block in either version because there aren't any long blocking calls (startContinuousSequence returns almost immediately - however `snapImage` will block the GUI loop in everything except the newest pymmcore:
<details>
 <summary>Test Script</summary>

```python
import sys
import time
from pathlib import Path

from loguru import logger

# use pymmcore to avoid version restrictions in pymmc+
from pymmcore import CMMCore
from qtpy.QtCore import QTimer
from qtpy.QtWidgets import (
    QApplication,
    QLabel,
    QMainWindow,
    QPushButton,
    QVBoxLayout,
    QWidget,
)
from superqt.utils import thread_worker

mm_pth = {
    "darwin": Path("/Applications/"),
    "win32": Path("C:/Program Files/"),
    "linux": Path("/usr/local/lib"),
}[sys.platform]
mm_pth = next(mm_pth.glob("[m,M]icro-[m,M]anager*"))


@thread_worker
def snap_worker(core):
    logger.info("pre snap")
    core.snapImage()
    logger.info("post snap")


@thread_worker()
def live_worker(core, frames=5):
    # avoids a GIL lock because this gets shifted way off into
    # a thread in the camera driver I think
    # https://github.com/micro-manager/mmCoreAndDevices/blob/a6bf0669764ddd4c64c9eafac12a54bfc2d03589/MMCore/MMCore.cpp#L2903
    logger.info("starting live")
    core.startContinuousSequenceAcquisition(500)
    for _ in range(frames):
        try:
            yield core.getLastImage()
        except (RuntimeError, IndexError):
            pass
        time.sleep(0.5)
    logger.info("done live")


class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.mmc = CMMCore()
        self.mmc.setDeviceAdapterSearchPaths([str(mm_pth)])
        self.mmc.loadSystemConfiguration(str((mm_pth / "MMConfig_demo.cfg").resolve()))
        self.mmc.setExposure(5000)  # 5 seconds

        self.counter = 0

        self.mmc.snapImage
        layout = QVBoxLayout()

        self.l = QLabel("Start")
        snap_button = QPushButton("Snap image!")
        snap_button.pressed.connect(self.snap)

        live_button = QPushButton("Go live")
        live_button.pressed.connect(self.live)

        layout.addWidget(self.l)
        layout.addWidget(snap_button)
        layout.addWidget(live_button)

        w = QWidget()
        w.setLayout(layout)

        self.setCentralWidget(w)

        self.show()

        self.timer = QTimer()
        self.timer.setInterval(100)
        self.timer.timeout.connect(self.recurring_timer)
        self.timer.start()

    def snap(self):
        snap_worker(self.mmc).start()

    def live(self):
        live_worker(self.mmc).start()

    def recurring_timer(self):
        self.counter += 1
        self.l.setText("Counter: %d" % self.counter)


app = QApplication(sys.argv)
window = MainWindow()
app.exec()
```
</details>

**Comment by ianhi** [comment]
I guess a downside of this is that  (i think?) we would lose the nice ability to connect to an already running mmcore instance from a new process (e.g. napari-micromanager starts one and then I connect from a jupyter notebook)

**Comment by ianhi** [comment]
> Or maybe I've misunderstood and the correct move is to keep pymmc+ as is and use the threading model in napari-micromanager?

I think there's a mix here. napari-micromanager should use threads in it's `snapImage` update, but here we need to implement locks for safety and use threading on `run_mda`

**Comment by ianhi** [comment]
fixed by #73
